pallets/rps/src/lib.rs

```diff
@@ -14,89 +13,427 @@ mod tests;
 #[cfg(feature = "runtime-benchmarks")]
 mod benchmarking;
 
+pub type ChallengeId = u64;
+
+pub type ChallengePlayHash = [u8; 8];
+
+#[derive(Clone, Encode, Decode, TypeInfo, MaxEncodedLen)]
+pub struct OpenChallenge<AccountId: PartialEq + Clone, Balance> {
```

Comment:
I like that `OpenChallenge`, `AcceptedChallenge` and `FinishedChallenge` are explicitly separated but they seem to share some fields together.

Can we combine them into one struct - something like:
```rust
pub struct Challenge<AccountId: PartialEq + Clone, Balance> {
	challenger: AccountId,
	rival: Option<AccountId>, // for opening a challenge
	bet_amount: Balance,
	winner: Option<AccountId>,
}
```

Reply:
Yeah, the reason I separated them is to make state transitions more explicit and reduce the amount of defensive code I would need to create to validate the state of the structs themselves.

But I can try changing it into a single struct If you prefer it like that.